### PR TITLE
feat(openchallenges): model input data types of challenges with EDAM

### LINKED ISSUE #2551: [Task] Model input data types of challenges with EDAM
### What product(s) is this story for?

OpenChallenges

### As a user, I want

_No response_

### Description

Refactor #2265 to model input data types as EDAM concepts

### Acceptance criteria

_No response_

### Tasks

- [ ] Update the Data Model
- [ ] Update the API model
- [ ] Update the challenge service
- [ ] Update the API client for Angular
- [ ] Update the web app

Out of scope:

- Index the new properties into ES

### Anything else?

_No response_

### Have you linked this story to a GitHub Project?

- [X] I have linked this story to a GitHub Project and set its metadata.

### PR DESCRIPTION
**Pull request by tschaffter**
Closes #2551

## Changelog

- model `Challenge.inputDataTypes` as an array of `EdamData` objects
- remove the endpoint used to explore the collection of custom input data types
- add README.md to the folder `src/main/resources/db` to share notes about how data are imported from CSV files

## Future Work

- Remove dummy input data type (another PR), keeping them in this PR for the sake of testing
- Create a facet to explore used EDAM input data types

### Search using the full-text search field

One challenge annotated with "Resource type" (`http://edamontology.org/data_0005`)

Searching with EDAM preferred name:

Screen Shot 2024-03-13 at 15 01 58

Searching with short EDAM class ID:

Screen Shot 2024-03-13 at 15 04 32

Searching with full EDAM class ID:

Screen Shot 2024-03-13 at 15 05 08

### `Challenge.inputDataTypes` is now an array of EDAM Data objects

```
http://localhost:8085/v1/challenges/1

{
  "id": 1,
  "slug": "network-topology-and-parameter-inference",
  "name": "Network Topology and Parameter Inference",
  "headline": "Optimize methods to estimate biology model parameters",
  "description": "Participants are asked to develop and/or apply optimization methods, including the selection of the most informative experiments, to accurately estimate parameters and predict outcomes of perturbations in Systems Biology models.",
  "doi": "",
  "status": "completed",
  "platform": {
    "id": 1,
    "slug": "synapse",
    "name": "Synapse"
  },
  "websiteUrl": "https://www.synapse.org/#!Synapse:syn2821735",
  "avatarUrl": "",
  "incentives": [
    "publication"
  ],
  "submissionTypes": [
    "prediction_file"
  ],
  "inputDataTypes": [
    {
      "classId": "http://edamontology.org/data_0005",
      "preferredLabel": "Resource type"
    }
  ],
  "categories": [],
  "startDate": "2012-06-01",
  "endDate": "2012-10-01",
  "starredCount": 2,
  "operation": null,
  "createdAt": "2023-11-15T22:40:15Z",
  "updatedAt": "2024-03-04T18:31:19Z"
}
```

### Example of challenge with input data type indexed in Elasticsearch

```
http://localhost:9200/openchallenges-challenge-000001/_search?q=(name:Network%20Topology%20and%20Parameter%20Inference)&size=1


```

### ES indexed values may be a single objects or an array

> [!NOTE]  
> What is describe below is not an issue. It is simply how ES index data. The property returned by the OC Challenge Service will still always be an array.

The property `Challenge.inputDataTypes` takes an array as value when the challenge has multiple types. However, this property takes the value of a single object when its has a single type. This property must always be an array.

```
"id","challenge_id","edam_concept_id","created_at"
"1","1","1","2023-06-23 00:00:00"

http://localhost:9200/openchallenges-challenge-000001/_search?q=(name:Network%20Topology%20and%20Parameter%20Inference)&size=1

          "input_data_types": {
            "class_id": "http://edamontology.org/data_0005",
            "preferred_label": "Resource type"
          },
```

vs

```
"id","challenge_id","edam_concept_id","created_at"
"1","1","1","2023-06-23 00:00:00"
"2","1","2","2023-06-23 00:00:00"

http://localhost:9200/openchallenges-challenge-000001/_search?q=(name:Network%20Topology%20and%20Parameter%20Inference)&size=1

          "input_data_types": [
            {
              "class_id": "http://edamontology.org/data_0005",
              "preferred_label": "Resource type"
            },
            {
              "class_id": "http://edamontology.org/data_0007",
              "preferred_label": "Tool"
            }
          ],
```

cc: @vpchung @mdsage1 @gaiaandreoletti

### CODE REVIEW

**Comment by tschaffter** [comment]
## Error after importing changes from merged EDAM Operation feature

```
2024-03-12 17:24:34 ERROR [main] o.h.s.e.r.spi.RootFailureCollector - HSEARCH000521: Hibernate Search encountered a failure during bootstrap; continuing for now to list all problems, but the process will ultimately be aborted.
Context: Hibernate ORM mapping, type 'org.sagebionetworks.openchallenges.challenge.service.model.entity.EdamOperationEntity'
Failure:
org.hibernate.search.util.common.SearchException: HSEARCH000588: Multiple entity types mapped to index 'openchallenges-edam-concept': 'EdamDataEntity', 'EdamOperationEntity'. Each indexed type must be mapped to its own, dedicated index.
        at org.hibernate.search.engine.common.impl.IndexManagerBuildingStateHolder$BackendInitialBuildState.createIndexManagerBuildingState(IndexManagerBuildingStateHolder.java:170)
        at org.hibernate.search.engine.common.impl.IndexManagerBuildingStateHolder.getIndexManagerBuildingState(IndexManagerBuildingStateHolder.java:84)
        at org.hibernate.search.engine.common.impl.MappedIndexManagerFactoryImpl.createMappedIndexManager(MappedIndexManagerFactoryImpl.java:29)
        at org.hibernate.search.mapper.pojo.mapping.building.impl.PojoMapper.mapIndexedType(PojoMapper.java:221)
        at org.hibernate.search.mapper.pojo.mapping.building.impl.PojoMapper.mapIndexedTypes(PojoMapper.java:199)
        at org.hibernate.search.engine.common.impl.SearchIntegrationBuilder$MappingBuildingState.mapIndexedTypes(SearchIntegrationBuilder.java:282)
        at org.hibernate.search.engine.common.impl.SearchIntegrationBuilder.prepareBuild(SearchIntegrationBuilder.java:180)
        at org.hibernate.search.mapper.orm.bootstrap.impl.HibernateSearchPreIntegrationService$NotBooted.doBootFirstPhase(HibernateSearchPreIntegrationService.java:246)
        at org.hibernate.search.mapper.orm.bootstrap.impl.HibernateOrmIntegrationBooterImpl.bootNow(HibernateOrmIntegrationBooterImpl.java:173)
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
        at java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)
        at org.hibernate.search.mapper.orm.bootstrap.impl.HibernateSearchSessionFactoryObserver.sessionFactoryCreated(HibernateSearchSessionFactoryObserver.java:41)
        at org.hibernate.internal.SessionFactoryObserverChain.sessionFactoryCreated(SessionFactoryObserverChain.java:35)
        at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:402)
        at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:471)
        at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.build(EntityManagerFactoryBuilderImpl.java:1498)
        at org.springframework.orm.jpa.vendor.SpringHibernateJpaPersistenceProvider.createContainerEntityManagerFactory(SpringHibernateJpaPersistenceProvider.java:58)
        at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:365)
        at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:409)
        at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.afterPropertiesSet(AbstractEntityManagerFactoryBean.java:396)
        at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.afterPropertiesSet(LocalContainerEntityManagerFactoryBean.java:341)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1863)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1800)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:233)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveNamedBean(DefaultListableBeanFactory.java:1284)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveNamedBean(DefaultListableBeanFactory.java:1245)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveBean(DefaultListableBeanFactory.java:494)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:349)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBean(DefaultListableBeanFactory.java:342)
        at org.springframework.orm.jpa.EntityManagerFactoryUtils.findEntityManagerFactory(EntityManagerFactoryUtils.java:121)
        at org.springframework.orm.jpa.EntityManagerFactoryAccessor.setBeanFactory(EntityManagerFactoryAccessor.java:155)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeAwareMethods(AbstractAutowireCapableBeanFactory.java:1826)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1791)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1391)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1311)
        at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:887)
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:791)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:541)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.addCandidateEntry(DefaultListableBeanFactory.java:1609)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1573)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveMultipleBeans(DefaultListableBeanFactory.java:1462)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1349)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1311)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredMethodElement.resolveMethodArguments(AutowiredAnnotationBeanPostProcessor.java:760)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredMethodElement.inject(AutowiredAnnotationBeanPostProcessor.java:720)
        at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:119)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessProperties(AutowiredAnnotationBeanPostProcessor.java:399)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1431)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:619)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:410)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1391)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1311)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredMethodElement.resolveMethodArguments(AutowiredAnnotationBeanPostProcessor.java:760)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredMethodElement.inject(AutowiredAnnotationBeanPostProcessor.java:720)
        at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:119)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessProperties(AutowiredAnnotationBeanPostProcessor.java:399)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1431)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:619)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:410)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.
```

**Review by gaiaandreoletti — APPROVED** [review]

File: apps/openchallenges/challenge-service/src/main/resources/db/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+## Challenge Service DB
+
+### CSV headers are ignored
+
+The headers of the CSV files are ignored by the REST API service. The headers are only included to provide additional information to developers.
```
Comment (vpchung) [inline apps/openchallenges/challenge-service/src/main/resources/db/README.md:5]: Good note.

File: apps/openchallenges/challenge-service/src/main/resources/db/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+Example:
+
+```csv
+"id","challenge_id","category"
+"1","55","benchmark"
+```
+
+### SQL `AUTO_INCREMENT` overwrites data from CSV files
+
+SQL columns annotated with `AUTO_INCREMENT` will ignore the column data imported from CSV files.
```
Comment (vpchung) [inline apps/openchallenges/challenge-service/src/main/resources/db/README.md:16]: Also, a good note 👍🏼

File: apps/openchallenges/db-update/update_db_csv.py [hunk 3]
```diff
@@ -216,18 +222,20 @@ def main(gc):
...
-               output_folder=CHALLENGE_FOLDER, print_row=True)
+    output_csv(
+        incentives, "incentives.csv", output_folder=CHALLENGE_FOLDER, print_row=True
+    )
+    output_csv(
+        sub_types,
+        "submission_types.csv",
+        output_folder=CHALLENGE_FOLDER,
+        print_row=True,
+    )
```
Comment (vpchung) [inline apps/openchallenges/db-update/update_db_csv.py:238]: Should we go ahead and update this dumping script to include the dump file for `input_data_type.csv`?  Or we can save it for another PR.

**Review by vpchung — APPROVED** [review]
Works on my end - thanks!

File: apps/openchallenges/db-update/update_db_csv.py (hunk 3 above)
Comment (tschaffter) [reply apps/openchallenges/db-update/update_db_csv.py:238]: @vpchung Could you update this file as part of the present PR?

Comment (vpchung) [reply apps/openchallenges/db-update/update_db_csv.py:241]: Done ✅

**Pull request merged**